Lib/fontmake/instantiator.py

```diff
@@ -0,0 +1,545 @@
+    glyph_new = font[name_new]
+
+    p = glyph_swap.getPointPen()
+    glyph_old.drawPoints(p)
+    glyph_swap.width = glyph_old.width
+
+    glyph_old.clear()
+    p = glyph_old.getPointPen()
+    glyph_new.drawPoints(p)
+    glyph_old.width = glyph_new.width
+
+    glyph_new.clear()
+    p = glyph_new.getPointPen()
+    glyph_swap.drawPoints(p)
+    glyph_new.width = glyph_swap.width
+
+    # 2. Remap components.
+    for g in font:
+        for c in g.components:
+            if c.baseGlyph == name_old:
+                c.baseGlyph = name_new
+            elif c.baseGlyph == name_new:
+                c.baseGlyph = name_old
+
+    # 3. Swap literal names in kerning.
+    kerning_new = {}
+    for first, second in font.kerning.keys():
+        value = font.kerning[(first, second)]
+        if first == name_old:
+            first = name_new
+        elif first == name_new:
+            first = name_old
+        if second == name_old:
+            second = name_new
+        elif second == name_new:
+            second = name_old
+        kerning_new[(first, second)] = value
+    font.kerning = kerning_new
+
+    # 4. Swap names in groups.
+    for group_name, group_members in font.groups.items():
+        group_members_new = []
+        for name in group_members:
+            if name == name_old:
+                group_members_new.append(name_new)
+            elif name == name_new:
+                group_members_new.append(name_old)
+            else:
+                group_members_new.append(name)
+        font.groups[group_name] = group_members_new
+
+
+@attr.s(auto_attribs=True, frozen=True, slots=True)
+class Variator:
+    """A middle-man class that ingests a mapping of normalized locations to
+    masters plus axis definitions and uses varLib to spit out interpolated
+    instances at specified normalized locations.
+
+    fontMath objects stand in for the actual master objects from the
+    UFO. Upon generating an instance, these objects have to be extracted
+    into an actual UFO object.
+    """
+
+    masters: List[FontMathObject]
+    location_to_master: Dict[Tuple[Tuple[str, float], ...], FontMathObject]
+    model: varLib.models.VariationModel
+
+    @classmethod
+    def from_masters(
+        cls, items: List[Tuple[Location, FontMathObject]], axis_order: List[str]
+    ):
+        masters = []
+        master_locations = []
+        location_to_master = {}
+        for normalized_location, master in items:
+            master_locations.append(normalized_location)
+            masters.append(master)
+            location_to_master[tuple(normalized_location.items())] = master
+        model = varLib.models.VariationModel(master_locations, axis_order)
+
+        return cls(masters, location_to_master, model)
+
+    def instance_at(self, normalized_location: Location) -> FontMathObject:
+        # Optimization: If a location matches a master, return the master instead of
+        # running through varLib.
```

Comment:
this is more than just a simple "optimization". It would allow to have "instances" for masters that are not interpolation-compatible. Think about the way in Glyphs.app you can only ever export instances, never masters as such; so even with a single master font, you always export its implicit "Regular" instance. Instances may carry additional data than the bare masters. With fontmake the `apply_instance_data` step is only done when you pass the `-i` (`--interpolate`) option. However one cannot use that option unless the masters are interpolatable, even if interpolation itself may not be required since all instances defined match the same location as their respetcive masters.
Would be nice to implement this feature in instantiator if it's not too much work. I think it's basically just a matter of identifying the master that correspond to a given instance's location, and returning a copy of its data.

Reply:
Already possible with the code :+1:

Reply:
I think this may solve another problem we currently have with a client project: we want to ship a VF and static fonts from the same sources. The VF needs a glyph replacement rule for `dollar` and `cent`, but we currently can't easily apply them to the statics (without also messing with our VTT dumps), so we include both and do a cmap swap hack. This would formalize the hack :)

Reply:
Mh, this reminds me of my question further up: when replacing glyphs in instances, should we swap or replace the glyphs in question?

Reply:
oh, I misunderstood the comment, I see that feature is already implemented :)
But I see you're returning the same master object, not a copy. Would that be a problem if following code assumes instanced objects are distinct from the masters'?

> when replacing glyphs in instances, should we swap or replace the glyphs in question?

what's the difference? isn't it the same thing?

Reply:
(adressed in latest pushes)